File: src/sites/BaseSite.ts
import { Message } from 'discord.js';
import ProcessResponse from './ProcessResponse';

abstract class BaseSite {
    identifier = 'Base';

    pattern = /base/i;

    color = 0x000000;

    match(message: string): IterableIterator<RegExpMatchArray> {
        return message.matchAll(this.pattern);
    }

    abstract process(
        match: RegExpMatchArray,
        source: Message | null
    ): Promise<ProcessResponse | false>;
}

export default BaseSite;

File: SaucyBot/Site/BaseSite.cs
using System.Reflection;
using System.Text.RegularExpressions;
using Discord;

namespace SaucyBot.Site;

public abstract class BaseSite : IBaseSite
{
    public string Identifier => GetType().GetCustomAttribute<SiteIdentifierAttribute>()?.Identifier
        ?? throw new InvalidOperationException($"Site type '{GetType().Name}' has no site identifier.");

    public virtual Regex Pattern { get; protected init; } = new(string.Empty);

    public virtual Color Color => Color.Default;

    public abstract Task<ProcessResponse?> Process(ProcessRequest request);
}
